### DISCUSSION TITLE
How do I create flat buttons and load an image onto it?

### DISCUSSION THREAD

**Discussion by M4SSD35TRUCT10N** [fltk-rs/fltk-rs#462] [Q&A] [answered]
In my "little" project I do want my buttons flat, with a divider in between and drawn with an image. I use the theme Gleam as my default theme.

The function `set_down_frame(FrameType::FlatBox)` doesn't seem to do anything when applied to a button.

My buttons are within a horizontal pack in a Group which itself is in a Tabs. The divider between the buttons would be eye candy and is not needed that much (I guess I can do it with an one/two pixel wide line/rectangle primitive).

I also tried to invoke function `image(<path to png>)` (as seen on the cheat sheet side for C/C++ - fltk) on a button but that doesn't seem to work.

What am I missing here? Sorry to ask such questions. I do feel like a beginner here who has never ever written any code. In fact I have never written an application in a systems language.

**Comment by MoAlyousef** [comment, marked as answer]
Hi
The set_down_frame sets the frame type when the button is clicked. You can try the set_frame method.
`mybutton.set_frame(FrameType::FlatBox);`
You can override the draw method to draw a line as a separator:
```rust
use fltk::*;

struct MyButton {
    but: button::Button,
}

impl MyButton {
    pub fn new(x: i32, y: i32, w: i32, h: i32, label: &str) -> Self {
        let mut but = button::Button::new(x, y, w, h, label);
        but.set_frame(FrameType::FlatBox);
        but.draw2(|b| {
            draw::set_draw_hex_color(0x464646);
            draw::set_line_style(draw::LineStyle::JoinRound, 2);
            draw::draw_line(
                b.x(),
                b.y() + b.height() + 5,
                b.x() + b.width(),
                b.y() + b.height() + 5,
            );
        });
        MyButton { but }
    }
}

impl std::ops::Deref for MyButton {
    type Target = button::Button;

    fn deref(&self) -> &Self::Target {
        &self.but
    }
}

impl std::ops::DerefMut for MyButton {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.but
    }
}

fn main() {
    let app = app::App::default().with_scheme(app::Scheme::Gleam);
    let mut wind = window::Window::new(100, 100, 400, 300, "Hello from rust");
    let mut frame = frame::Frame::new(0, 0, 400, 200, "");
    let mut but = MyButton::new(160, 210, 80, 40, "Click me!");

    wind.end();
    wind.show();

    but.set_callback(move || frame.set_label("Hello world"));
    app.run().unwrap();
}
```
Or you can use thin frames with of RFlatBox (rounded flat box) as separators.
As for setting the image, after loading the image using one of the image initializers, you'll have to call mybutton.set_image(Some(img));
```rust
    let mut image = SharedImage::load("path/to/image").unwrap();
    image.scale(200, 200, true, true);
    but.set_image(Some(image));
```
Or you can draw the image directly inside the button:
```rust
    but.draw2(|b| {
        /* previous code */
        let mut image = SharedImage::load("path/to/image").unwrap();
        image.scale(b.width(), b.height(), true, true);
        image.draw(b.x(), b.h(), b.width(), b.height());
    });
```

**Reply by M4SSD35TRUCT10N** [reply]
Thank you for your help! Again.

It was very late (2300+) and just as I switched off the monitor I saw within the fltk-derive (if I remember it correctly) sources the function `set_frame`, but was too tired to try it. But I guess I found my fault: I didn't put a frame first, so my buttons (in a `Pack`) are kinda free floatin'.

As for the image I assume I always have to do this one way or another:
```
let mut image = SharedImage::load("path/to/image").unwrap();
image.scale(200, 200, true, true); // my milage may vary here
but.set_image(Some(image));
```

Why I can't find the function 'draw2' in the docs? There's only `draw`? Obviously it works but creates an error.
![image](https://user-images.githubusercontent.com/13597659/106245905-90e5ac80-620d-11eb-9939-340a50b47d70.png)

**Comment by MoAlyousef** [comment]
The scale is only needed if the image itself is large. If you have small icons and their size is perfect, you don’t need to call scale. 
Check the docs version of fltk, sometimes google sends you to an outdated version. 

It seems LineStyle support is platform-dependent.
